[FLINK-7442] Add option for using child-first classloader for loading user code

aljoscha opened the pull request:
This is an alternative to #4554 that does not make the client class loader configurable.

## What is the purpose of the change

This PR introduces a new core option (`classloader.resolve-order: child-first`) that allows using a child-first class loader for user code. The default is still to use a parent-first class loader.

This also does a minor refactoring in the way the blob manager retrieves the cleanup interval. It's now also read from the `Configuration`, since we already have the `Configuration` for the class loader settings.

## Brief change log

 - Introduce new option
 - Pass `Configuration` thought to all places where we previously created a user class loader
 - Instantiate correct class loader based on config

## Verifying this change

This PR introduces new end-to-end tests that verify the new feature in a complete Flink workflow, including starting the program using `bin/flink run`.

## Does this pull request potentially affect one of the following parts:


This affects class loader, which is quite important to get right.

## Documentation

 - the new flag is documented in the config documentation

StephanEwen commented:
Thanks, this looks better to me!

I would suggest to not pass the configuration into the library cache manager.
I think passing configuration objects into specialized / dedicated components is an anti-pattern. It makes testing complicated, signatures inexplicit, etc.

For the second task of the classloader, loading resources: Do we have a test that validates the resource resolution? I think this passes the tests of the previous pull request because it behaves for resources like a *parent-first classloader*, which seems inconsistent. Admittedly the existing implementation of the child-first classloader was for resources a child-only classloader, which was also not correct.

aljoscha commented:
@StephanEwen I updated the PR. Turns out you also have to override the resource-related methods to change the resolution order. I added checks for that to the end-to-end test.

aljoscha commented:
@StephanEwen I finally ironed out the last kinks. I think this is now ready for a final review.

StephanEwen commented:
Did a side-by-side review with @aljoscha and directly fixed issues.
This is in very good shape from my side now.

+1 to merge this
